app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java

```diff
@@ -0,0 +1,310 @@
+     * Activity will quit upon exiting the dialog.
+     *
+     * @param msg String message that should be shown on the alert
+     */
+    private void showAlertDialog(String msg) {
+        AlertDialog.Builder alert = new AlertDialog.Builder(this);
+        alert.setTitle(msg);
+        alert.setCancelable(false);
+        alert.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
+
+            public void onClick(DialogInterface dialog, int which) {
+                dialog.dismiss();
+                finish();
+            }
+        });
+
+        alert.show();
+    }
+
+    /**
+     * Prepares a WebView of the html document generated by this activity, which can then be
+     * printed by the Android print framework
+     *
+     * Source: https://developer.android.com/training/printing/html-docs.html
+     */
+    private void doHtmlPrint() {
+        // Create a WebView object specifically for printing
+        WebView webView = new WebView(this);
+        webView.setWebViewClient(new WebViewClient() {
+
+            public boolean shouldOverrideUrlLoading(WebView view, String url) {
+                return false;
+            }
+
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                createWebPrintJob(view);
+                mWebView = null;
+            }
+        });
+        try {
+            String htmlDocString = TemplatePrinterUtils.readStringFromFile(outputPath);
+            webView.loadDataWithBaseURL(null, htmlDocString, "text/HTML", "UTF-8", null);
+            // Keep reference to WebView object until PrintDocumentAdapter is passed to PrintManager
+            mWebView = webView;
+        } catch (IOException e) {
+            showErrorDialog(Localization.get("print.io.error"));
+        }
+
+    }
+
+    /**
+     * Starts a print job for the given WebView
+     *
+     * Source: https://developer.android.com/training/printing/html-docs.html
+     *
+     * @param v the WebView to be printed
+     */
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    private void createWebPrintJob(WebView v) {
+
+        // Get a PrintManager instance
+        PrintManager printManager = (PrintManager) getSystemService(Context.PRINT_SERVICE);
+
+        // Get a print adapter instance
+        PrintDocumentAdapter printAdapter = new PrintDocumentAdapterWrapper(v.createPrintDocumentAdapter());
+
+        // Create a print job with name and adapter instance
+        printJob = printManager.print(jobName, printAdapter, new PrintAttributes.Builder().build());
+    }
+
+    /**
+     * A wrapper for the default print document adapter generated for a web view, to enable
+     * implementation of a custom callback when onFinish() is called
+     *
+     * Source: http://stackoverflow.com/questions/30742051/android-printmanager-get-callback
+     */
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    class PrintDocumentAdapterWrapper extends PrintDocumentAdapter {
```

Comment:
I purposefully left this here for 2 reasons: 1) It really has no use outside of this activity. 2) If it were in a separate file it would need to take in the TemplatePrinterActivity as a field, since it calls showAlertDialog, which seems unnecessarily messier.